rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,516 @@
+        options: DbReaderOptions
+    ) -> Result<Self, SlateDBError> {...}
+
+    /// Read a key an return the read value, if any.
+    pub async fn get(&self, key: &[u8]) -> Result<Option<Bytes>, SlateDBError> {...}
+}
+
+pub struct GarbageCollectorOptions {
+    /// Defines a grace period for deletion of a db. A db for which deletion is handled by the GC
+    /// will not be deleted from object storage until this duration has passed after it's deletion
+    /// timestamp.
+    pub db_delete_grace: Duration,
+}
+```
+
+Finally, we’ll also allow users to create/refresh/delete/list checkpoints using the admin CLI:
+
+```bash
+$ slatedb create-checkpoint --help
+Create a new checkpoint pointing to the database's current state
+
+Usage: slatedb --path <PATH> create-checkpoint [OPTIONS]
+
+Options:
+  -l, --lifetime <LIFETIME>  Optionally specify a lifetime for the created checkpoint. You can
+    specify the lifetime in a human-friendly format that uses years/days/min/s, e.g. "7days 30min 10s".
+    The checkpoint's expiry time will be set to the current wallclock time plus the specified
+    lifetime. If the lifetime is not specified, then the checkpoint is set with no expiry and
+    must be explicitly removed
+  -s, --source <UUID> Optionally specify an existing checkpoint to use as the base for the newly
+    created checkpoint. If not provided then the checkpoint will be taken against the latest
+    manifest.
+
+$ slatedb refresh-checkpoint --help
+Refresh an existing checkpoint's expiry time. This command will look for an existing checkpoint
+and update its expiry time using the specified lifetime
+
+Usage: slatedb --path <PATH> refresh-checkpoint [OPTIONS] --id <ID>
+
+Options:
+  -i, --id <ID>              The UUID of the checkpoint (e.g. 01740ee5-6459-44af-9a45-85deb6e468e3)
+  -l, --lifetime <LIFETIME>  Optionally specify a new lifetime for the checkpoint. You can specify
+    the lifetime in a human-friendly format that uses years/days/min/s, e.g. "7days 30min 10s". The
+    checkpoint's expiry time will be set to the current wallclock time plus the specified lifetime.
+    If the lifetime is not specified, then the checkpoint is updated with no expiry and must be
+    explicitly removed
+
+$ slatedb delete-checkpoint --help
+Delete an existing checkpoint
+
+Usage: slatedb --path <PATH> delete-checkpoint --id <ID>
+
+Options:
+  -i, --id <ID>  The UUID of the checkpoint (e.g. 01740ee5-6459-44af-9a45-85deb6e468e3)
+
+$ slatedb list-checkpoints --help
+List the current checkpoints of the db.
+
+Usage: slatedb --path <PATH> list-checkpoints
+
+
+```
+
+### Creating a Checkpoint
+
+Checkpoints can be created using the `create_checkpoint` API, the `create-checkpoint` CLI command, or by the
+writer/reader clients as part of their normal operations (see below sections).
+
+Checkpoints can be taken from the current manifest, or from an existing checkpoint. Checkpoints cannot be created from
+arbitrary manifest versions.
```

Comment:
I got the answer to the other comment here. Can you please add why we have this restriction?

Reply:
I think we have a race condition between the checkpointer and the garbage collector otherwise. A checkpointer could see that manifest N still exists, read the latest manifest, add a new checkpoint for manifest N, and write the new manifest out successfully. But manifest N could be deleted at any point in that process by the GC. I suppose there are a few things we could do here:

1. IIRC, we have a configurable deletion age for manifests in the GC. So, the checkpointer could be made to only checkpoint manifests that are within some shorter time range.
2. Even if the manifest is deleted after the checkpoint is created, we could add a subsequent check to validate that the manifest still exists.
3. We could introduce some heuristic like the last M manifests are always kept regardless of their age or something.

There are probably other tweaks as well. Not sure if it's worth it. But this is my recollection, anyway.

Reply:
Thanks!
Things like "as of" semantics might require us to checkpoint old manifest, but it can also be solved by increasing the gc age to the maximum "as of" duration that we want to support. 
Anything more than current manifest is not needed until we find some real usecase.